### DISCUSSION TITLE
SQS consumer stops polling after a few days

### DISCUSSION THREAD

**Discussion by switz** [bbc/sqs-consumer#381] [Help] [answered]
I've noticed after a few days of uptime, my sqs consumers will fail to get any new data until I restart the app server. This wasn't an issue on older versions, but I believe it began after 6.0/7.0.

I am logging errors and processing errors, but not seeing anything in my logs. I'll just notice after a day that things got backed up. The second I reboot the server, every standing message gets picked up instantly.

Any thoughts? I wish I had more info, but I'm not sure how to debug this further than logging the errors below.

```
app.on('error', (err) => {
  console.error(err);
  winston.error('main sqs error', { err: JSON.stringify(err && err.message) });
});

app.on('processing_error', (err) =>
  winston.error('main sqs processing error', { err: JSON.stringify(err && err.message) })
);
```

**Comment by nicholasgriffintn** [comment]
Are you logging unhandled errors in your app?

This is quite likely to be an external resource outside of SQS Consumer, realistically, it's not doing anything all that complex.

**Reply by switz** [reply]
yeah, and restarting the app on those, also `unhandledRejection`:

```
  process.on('uncaughtException', (err) => {
    console.error(err);

    logger.error('Uncaught Exception thrown', err);
    process.exit(1);
  });
```

I guess I can add more logging around that code path, maybe I'm missing something. I'll look into it further, thanks!

**Reply by nicholasgriffintn** [reply]
Cool, yeah I would guess that something is failing at some point. SQS Consumer should be emitting any errors, but you can also enable debug for even more logging from our side. (env var `DEBUG=sqs-consumer`).

Feel free to update if you find anything out.

**Reply by switz** [reply]
So I was able to catch this happening with debug enabled:

![image](https://github.com/bbc/sqs-consumer/assets/990166/824e2a64-0c89-4af8-b4ce-60b95987feeb)

the processing practice logs are mine – the second I rebooted the server it started picking up messages again.

Any ideas how I can debug this further? This is happening across multiple sqs consumers across multiple app servers.

**Reply by switz** [reply]
Here's all of my code:

My only guess is that the fetch request is infinitely timing out? but I thought there was a timeout error that would catch that? `timeout_error` Or do I have to manually specify a `handleMessageTimeout`?

```javascript
import { Consumer } from 'sqs-consumer';
import fetch from 'isomorphic-fetch';
import { SQSClient } from '@aws-sdk/client-sqs';

import winston from './logger.js';

const PORT = process.env.PORT || 3000;

const app = Consumer.create({
  queueUrl: URL,
  batchSize: 10,
  sqs: new SQSClient({
    region: 'us-east-1',
  }),
  handleMessageBatch: (messages): Promise<any> => {
    return Promise.all(
      messages.map((message) => {
        if (!message?.Body) return message;
        try {
          const body = JSON.parse(message.Body);

          return fetch(`http://127.0.0.1:${PORT}/api/v1/${body.method}`, {
            method: 'POST',
            headers: {
              'Content-Type': 'application/json',
            },
            body: JSON.stringify(body),
          }).then(() => message);
        } catch (error) {
          const err = error;
          winston.error('sqs error caught', { err, message });
          return message;
        }
      })
    );
  },
});

app.on('error', (err) => {
  console.error(err);
  winston.error('main sqs error', { err: JSON.stringify(err && err.message) });
});

app.on('processing_error', (err) =>
  winston.error('main sqs processing error', { err: JSON.stringify(err && err.message) })
);

if (process.env.PROCESSOR) {
  app.start();
}
```

**Reply by nicholasgriffintn** [reply]
Hmmm, timeout_error is a different listener, I do wonder if an error from inside handleMessageBatch is not getting caught as a timeout error on the consumer's side shouldn't stop the service, it would just carry on as normal.

Are the consumer logs coming from the same process as the other logs? If it's an uncaught error, it's possible the process might have been stopped by node, it's either that or the consumer is just infinitely waiting to process the message, that's all I can really think of at the moment.

If messages are processed when the consumer is rebooted, it does suggest that it got stuck on a message.

**Reply by switz** [reply]
Yeah they’re the same process. Does handleMessageTimeout force a message to process or cancel in a certain amount of ms? Would specifying that fix a never ending fetch? Or Maybe I can find a fetch implementation with a timeout.

**Reply by nicholasgriffintn** [reply, marked as answer]
Yeah if `handleMessageTimeout` isn't defined, I don't think the consumer will ever timeout, it will just forever wait for your handler to complete, it's possible it might be that if you don't have a setting defined.

It did get me thinking that we should debug while messages are processing, which I added a ticket for here: https://github.com/bbc/sqs-consumer/issues/399

**Reply by switz** [reply]
Cool, I added that flag so we’ll see if it triggers. Extra debugging would be nice too! Thanks for the quick and cogent replies!

**Reply by nicholasgriffintn** [reply]
NP, let me know if it does the job or not.
